Lpm documentation update (#76)
TexasInstruments/processor-sdk-doc: source/linux/Foundational_Components/Power_Management/pm_low_power_modes.rst

         root@am62axx-evm:~# echo disabled > /sys/class/rtc/rtc0/device/power/wakeup
         root@am62axx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f900000.dwc3-usb/power/wakeup
         root@am62axx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f910000.dwc3-usb/power/wakeup

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

      .. code-block:: console

         root@am62pxx-evm:~# echo disabled > /sys/class/rtc/rtc0/device/power/wakeup
         root@am62pxx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f900000.usb/power/wakeup
         root@am62pxx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f910000.usb/power/wakeup

   .. rubric:: Then, configure PMIC register bit to turn off only selected rails for this mode.

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

      .. code-block:: console

         root@am62axx-evm:~# i2cset -f -y -m 0xFF -r -a 0 0x48 0x86 0x1

      The register write has been done to enable PMIC to enter `PMIC S2R <https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.ti.com/lit/pdf/slvucm3&ved=2ahUKEwjknrL864-KAxVocGwGHdJZFPwQFnoECBgQAQ&usg=AOvVaw1y0evgQxxy9YOF7l5Wzaef>`_ .

> 

Reviewer: fix the url to something like https://www.ti.com/lit/ug/slvucm3/slvucm3.pdf

Reviewer: can you also update the source/devices/<device>/linux/Release_Specific_Release_Notes.rst , Whats-new section hyperlinking to the key changes

Author: Updated. Thanks

#. Deep Sleep
#. MCU Only
#. Partial I/O

> #. I/O Only Plus DDR

Reviewer 2: I am wondering if we should sort these from lowest latency to highest latency just to be consistent?

Reviewer 2: ```suggestion
#. MCU Only
#. Deep Sleep
#. I/O Only Plus DDR
#. Partial I/O
```

Reviewer 3: I think this is a good idea if we explain that order in the above paragraph.

Author: We can list the ordering from lowest latency to highest latency separately but the following sections where these modes are explained are interdependent. e.g
MCU Only = Deep sleep + MCU core
I/O DDR = Partial I/O + DDR 
Will have to rewrite the entire section.
So, adding a separate table

Reviewer 2: @akashdeep-ti  please can you ping once you add this separate table?
Thanks.

Author: Updated in order of power consumption

MCU Only
********

.. _pm_mcu_only:

> 

Reviewer 3: `pm_wakeup_sources.rst` still contains a ref to this tag.

Reviewer 2: Good catch @StaticRocket !

Author: Updated

      devices that are mostly idle or low activity most of the time with the full context being saved.
   #. Better efficiency: I/O Only Plus DDR mode can help to improve the efficiency of embedded devices by
      reducing the amount of time that the processor is idle. This is because the processor can
      be kept in a low-power state when it is not needed.

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

      .. important:: Jumper J9 should be connected on SK to enable system to enter I/O Only plus DDR mode.

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

      .. important:: Jumper J12 should be connected on SK to enable system to enter I/O Only plus DDR mode.

   The wakeup sources that can be used to wake the system from I/O Only Plus DDR are
   mcu_uart0, mcu_mcan0, mcu_mcan1 and wkup_uart0. After Linux boots, direct register
   writes can be used to enable wakeup.

   .. rubric:: Following commands set the wakeup EN bit, enable receive for pad in PADCONFIG register and can
               be used to enable wakeup from mcu_mcan0, mcu_mcan1, mcu_uart0 and wkup_uart0 pins respectively.

   .. code-block:: console

>       root@<machine>:~# devmem2 0x4084038 0x20050000  # MCU_PADCONFIG14 for mcu_mcan0

Reviewer 4: This should be done via DT + pinctrl_pm_select_sleep_state() API in driver. Can we share a diff for the same instead. Note that  /dev/mem and thus devmem2 may be disabled in security hardened kernels. So this wont work always

Author: There are more changes required to enable the wakeup using dt, so added a note in documentation about this workaround.

   by a single interface signal (PMIC_LPM_EN signal) with PMIC register programming.

   The benefits of using I/O Only plus DDR in embedded devices:

   #. Lowest power consumption: I/O Only Plus DDR mode can save a significant amount of power, especially in battery-powered
      devices that are mostly idle or low activity most of the time with the full context being saved.
   #. Better efficiency: I/O Only Plus DDR mode can help to improve the efficiency of embedded devices by
      reducing the amount of time that the processor is idle. This is because the processor can
      be kept in a low-power state when it is not needed.

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

      .. important:: Jumper J9 should be connected on SK to enable system to enter I/O Only plus DDR mode.

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

      .. important:: Jumper J12 should be connected on SK to enable system to enter I/O Only plus DDR mode.

   The wakeup sources that can be used to wake the system from I/O Only Plus DDR are
   mcu_uart0, mcu_mcan0, mcu_mcan1 and wkup_uart0. After Linux boots, direct register
   writes can be used to enable wakeup.

>    .. rubric:: Following commands can be used to enable wakeup from mcu_mcan0, mcu_mcan1,

Reviewer 2: ```suggestion
   .. rubric:: Following commands set the wakeup EN bit in PADCONFIG register and can be used to enable wakeup from mcu_mcan0, mcu_mcan1, ..... pins
```

Author: Thanks for the suggestion. Applied

               mcu_uart0 and wkup_uart0 respectively.

   .. code-block:: console

      root@<machine>:~# devmem2 0x4084038 0x20050000
      root@<machine>:~# devmem2 0x4084040 0x20050000
      root@<machine>:~# devmem2 0x4084014 0x20050000

>       root@<machine>:~# devmem2 0x4084024 0x20050000

Reviewer 2: It would be better in my opinion if we could add some explanation on these commands below, atleast giving a reference to the PADCONFs being used and what bits we're setting in them.
For eg.

```suggestion
      root@<machine>:~# devmem2 0x4084024 0x20050000 # PADCONFIGN
```
This would help consumers make sense of what is happening here. Otherwise it's very difficult for someone to know why we are doing raw reg writes from userspace.

   .. note::

      Atleast one of the wakeup sources listed above must be enabled to wakeup from I/O Only Plus DDR mode.

   .. rubric:: To enter I/O Only Plus DDR mode, first disable wakeup from RTC, USB0 and USB1 as these wakeup
               sources are not supported for this mode.

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

      .. code-block:: console

         root@am62axx-evm:~# echo disabled > /sys/class/rtc/rtc0/device/power/wakeup
         root@am62axx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f900000.dwc3-usb/power/wakeup
         root@am62axx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f910000.dwc3-usb/power/wakeup

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

      .. code-block:: console

>          root@<machine>:~# echo disabled > /sys/class/rtc/rtc0/device/power/wakeup

Reviewer 5: Better to have a single convention for the entire section.

- If it's a common section, use \<machine\>
- if it's under ifconfig, use the actual yocto machine name

Since it's under CONFIG_part_variant in ('AM62PX')
```suggestion
         root@<am62pxx-evm>:~# echo disabled > /sys/class/rtc/rtc0/device/power/wakeup
```

Reviewer 3: I concur. Would you mind adding that to the contribution guide @cshilwant ?

Author: Updated. Thanks for pointing this out

         root@am62pxx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f900000.usb/power/wakeup

>          root@am62pxx-evm:~# echo disabled > /sys/devices/platform/bus@f0000/f910000.usb/power/wakeup

Reviewer 2: I don't see a difference between 62P and 62A commands here, can we not unify this then?
just make it as ``root@<machine>`` like in other places.

Author: The command to disable wakeup from USB0 and USB1 is different, so kept all wakeup disables together

Reviewer 5: @DhruvaG2000 dwc3-usb vs usb is the delta

Reviewer 2: oh okay misread it, thanks. Keep as is then.

   .. rubric:: Then, configure PMIC register bit to turn off only selected rails for this mode.

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

      .. code-block:: console

         root@am62axx-evm:~# i2cset -f -y -m 0xFF -r -a 0 0x48 0x86 0x1

      The register write has been done to enable PMIC to enter `PMIC S2R <https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.ti.com/lit/pdf/slvucm3&ved=2ahUKEwjknrL864-KAxVocGwGHdJZFPwQFnoECBgQAQ&usg=AOvVaw1y0evgQxxy9YOF7l5Wzaef>`_ .

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

      .. code-block:: console

>          root@am62axx-evm:~# i2cset -f -y -m 0xFF -r -a 0 0x48 0x86 0x2

Reviewer 5: ```suggestion
         root@am62pxx-evm:~# i2cset -f -y -m 0xFF -r -a 0 0x48 0x86 0x2
```

Since it's under CONFIG_part_variant in ('AM62PX')

Author: Agree. Updated. Thanks

   .. rubric:: Now, the SoC can be suspended using the following command:

   .. code-block:: console

         root@<machine>:~# echo mem > /sys/power/state
         [   26.132900] PM: suspend entry (deep)
         [   26.136759] Filesystems sync: 0.000 seconds
         [   26.151748] Freezing user space processes
         [   26.157256] Freezing user space processes completed (elapsed 0.001 seconds)
         [   26.164239] OOM killer disabled.
         [   26.167469] Freezing remaining freezable tasks
         [   26.173168] Freezing remaining freezable tasks completed (elapsed 0.001 seconds)
         [   26.180624] printk: Suspending console(s) (use no_console_suspend to debug)

>    This partially indicates that linux has finished it's I/O Only plus DDR sequence.

Reviewer 2: I think think this can be slightly misleading, the above log tells me nothing special that we are entering IO+DDR. This is the same for any suspend-to-mem scenario right?

Author: Exactly, that's why we added LED representation at the end

Reviewer 2: What I meant was let's reword this sentence to not say "linux has finished" rather say:
```suggestion
   This indicates that the device has partially completed the I/O Only plus DDR entry sequence.
```

>          [   26.180624] printk: Suspending console(s) (use no_console_suspend to debug)

Reviewer 3: Too much indentation

Author: corrected

   .. ifconfig:: CONFIG_part_variant in ('AM62AX')

         For further confirmation, one can take a look at the on board LED LD2 on the SK
         (LED should turn off).

   .. ifconfig:: CONFIG_part_variant in ('AM62PX')

         For further confirmation, one can take a look at the on board LED LD1 on the SK
         (LED should turn off).

   The system has entered I/O Only plus DDR and can be woken up either with an

>    activity on the I/O pin programmed for wakeup or key press on wkup_uart0 (third serial port /dev/ttyUSB2) or

Reviewer 5: ```suggestion
   activity on the I/O pin programmed for wakeup or key press on wkup_uart0 (third serial port :file:`/dev/ttyUSB2`) or
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

Author: Thanks. Updated
